manticore/ethereum.py

```diff
@@ -161,6 +163,19 @@ class Detector(Plugin):
     def name(self):
         return self.__class__.__name__.split('.')[-1]
 
+    @staticmethod
+    def is_in_dispatcher(state):
+        """
+        :param state: current state
+        :return: whether the current execution is in a dispatcher function or not
+        """
+        # The first 4 bytes are keccak256 of the function signature that we call with a transaction
+        # The Solidity dispatcher function takes it and dispatches the execution based on that
+        # So: if we are in the dispatcher, one of the bytes will effectively have more than one solutions
+
+        # TODO / FIXME: Benchmark/check if doing it byte by byte and also returning the bytes is faster
+        return len(state.solve_n(state.platform.current_transaction.data[:4], 2)) > 1
```

Comment:
heuristic could be improved..
- this works when the tx input starts having several possible solutions. 
- not differentiated from the fallback func. 
- the result may be cached globally. (Address, pc) -> dispatcher/funcname/fallback

Uh.. if there is SolidityMetadata associated you could use that too.

Reply:
@feliam 
> this works when the tx input starts having several possible solutions.
Hmm... So probably the only edge case is situation where contract have no other functions than a fallback one which and with current approach we won't detect dispatcher properly, is that right?

Do you have any idea how to distinguish `fallback` vs `dispatch`?

> the result may be cached globally. (Address, pc) -> dispatcher/funcname/fallback

Do you think it is okay to map `(address,pc)` to `FuncInfo` such as:
```python
class FuncType:
    DISPATCHER = 0
    FALLBACK = 1
    USER_FUNC = 2
    # maybe also constructor? but its not really used in transactions, so maybe not?

FuncInfo = namedtuple('FuncInfo', 'type, signature')
```

Reply:
Idea to differentiate? Not much.
- If you have access to the metadata you can try to parse the output from an  metadata.get_source_for(evm_offset).
- Use the compiler in other way
- Number of instructions executed could help in the heuristics otherwise 

As for the mapping to functype look out for the runtime vs. init code. Maybe you need to add a flag there to decide constructor/runtime. Other than that whatever works. but do it on the global manticore.context dict.